subsys/bluetooth/services/ots/ots.c

```diff
@@ -78,6 +78,50 @@ static ssize_t ots_obj_name_read(struct bt_conn *conn,
 				 strlen(ots->cur_obj->metadata.name));
 }
 
+#if CONFIG_BT_OTS_OBJ_NAME_WRITE_SUPPORT
+ssize_t ots_obj_name_write(struct bt_conn *conn,
+			   const struct bt_gatt_attr *attr,
+			   const void *buf, uint16_t len,
+			   uint16_t offset, uint8_t flags)
+{
+	struct bt_ots *ots = (struct bt_ots *) attr->user_data;
+	struct bt_gatt_ots_object *obj = NULL;
+	int rc = 0;
+	char name[CONFIG_BT_OTS_OBJ_MAX_NAME_LEN + 1];
+
+	LOG_DBG("OTS Object Name GATT Write Operation");
+
+	if (!ots->cur_obj) {
+		LOG_DBG("No Current Object selected in OTS!");
+		return BT_GATT_ERR(BT_GATT_OTS_OBJECT_NOT_SELECTED);
+	}
+
+	if (offset + len > CONFIG_BT_OTS_OBJ_MAX_NAME_LEN) {
+		LOG_DBG("Object name is too long!");
+		return BT_GATT_ERR(BT_GATT_OTS_WRITE_REQUEST_REJECTED);
+	}
+
+	/* Construct a temporary name for duplication detection */
+	strcpy(name, ots->cur_obj->metadata.name);
+	memcpy(name + offset, buf, len);
+	name[offset + len] = '\0';
+
+	rc = bt_gatt_ots_obj_manager_first_obj_get(ots->obj_manager, &obj);
+	while (rc == 0) {
+		if (obj != ots->cur_obj && strcmp(name, obj->metadata.name) == 0) {
+			LOG_DBG("Object name is duplicated!");
+			return BT_GATT_ERR(BT_GATT_OTS_OBJECT_NAME_ALREADY_EXISTS);
+		}
+		rc = bt_gatt_ots_obj_manager_next_obj_get(ots->obj_manager, obj, &obj);
+	}
+
+	/* Update real object name after no duplicate detected */
+	strcpy(ots->cur_obj->metadata.name, name);
```

Comment:
This should work on paper, but since we don't reject if offset > 0, then we allow write long. 

This means that if we have an object with the name `"Object with a very long name that do not fit in the MTU"`, and a client attempts to write the same name but with write long, then the client may write 
`"Object with a very long name that do not fit"` in the first write command (which passes the above since it's not a duplicate), and then attempt to add the remaining `" in the MTU"` which would cause it to be a duplicate, so we'd reject then, but we have already overwritten the current name with a part of the name. 

So in the above case we change the value of the name (which is stored in the application) with an incomplete write, while also not notifying the application about this change, both of which seem non-ideal. I don't have a solution for it (and maybe this is good enough for now), but otherwise I would suggest to either reject long writes for now (i.e. reject if offset > 0), and/or add a callback to the application that the name has changed.

Reply:
Good point about the long write. I think I will make a change to reject long writes.

As for your other point about the application not knowing about this update, I am not happy about this either. I was trying to start a conversation about this in my comment [here](https://github.com/zephyrproject-rtos/zephyr/pull/35996#issuecomment-855457722). I would like to hear your thoughts on this.

Maybe for now it is good enough that we don't tell the application and force the application to hold the name in memory as long as we reject long writes.

Reply:
I think adding a callback to the application would be the way to go, and should be fairly simple. 

The question is then if we should

1. Request the application to store the new name, which it can reject
2. Tell the application that OTS stored the new name, which it cannot reject (similar to what you added here). 

Since the text in the spec is: 
> If the Object Name characteristic is writable, the Server **shall allow** new values to be written to this characteristic by a Client.

Then I am leaning towards number 2, as there are also no application error codes that would fit a "No I don't want that name" response.